Add examples and tests for combining static/class method with async execution

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40619 Add examples and tests for combining static/class method with async execution**

Differential Revision: [D22258407](https://our.internmc.facebook.com/intern/diff/D22258407)

mrshenli committed:
Add examples and tests for combining static/class method with async execution

[ghstack-poisoned]

mrshenli committed:
Update on "Add examples and tests for combining static/class method with async execution"




[ghstack-poisoned]

mrshenli committed:
Update on "Add examples and tests for combining static/class method with async execution"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/functions.py:
```diff
@@ -73,12 +81,54 @@ def async_execution(fn):
+        >>> class AsyncExecutionClass:
+        >>>
+        >>>     @rpc.functions.async_execution
+        >>>     @staticmethod
```
In this case the decorator `rpc.functions.async` is outer, but it looks like the doc says it needs to be inner for static methods?

mrshenli replied:
This is a vestige from debug. let me update that. Thanks for spotting this.

mrshenli committed:
Update on "Add examples and tests for combining static/class method with async execution"


Differential Revision: [D22258407](https://our.internmc.facebook.com/intern/diff/D22258407)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks for clarifying the earlier comment I had! LGTM
